From bdb587699b4c30f9c541f0eb3b2028dc3345aad9 Mon Sep 17 00:00:00 2001
From: Zidan Wang <zidan.wang@freescale.com>
Date: Mon, 19 Jan 2015 11:26:41 +0800
Subject: [PATCH] MLK-10113: ASoC: fsl_sai: Bclk devision value must be a even
 number between 2-512

According to the RM of sai, bclk DIV is 8 bit, and the devision value is
(DIV+1)*2. So bclk devision value must be a even number between 2-512.

When find the best clock source, stop judging the others.

Signed-off-by: Zidan Wang <zidan.wang@freescale.com>
---
 sound/soc/fsl/fsl_sai.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/sound/soc/fsl/fsl_sai.c b/sound/soc/fsl/fsl_sai.c
index 4e90be3..c433728 100644
--- a/sound/soc/fsl/fsl_sai.c
+++ b/sound/soc/fsl/fsl_sai.c
@@ -346,8 +346,9 @@ static int fsl_sai_set_bclk(struct snd_soc_dai *dai, bool tx, u32 freq)
 		dev_dbg(dai->dev, "ratio %d for freq %dHz based on clock %ldHz\n",
 				ratio, freq, clk_rate);
 
-		ratio /= 2;
-		if (ratio == 0 && ratio > 256)
+		if (ratio % 2 == 0 && ratio >= 2 && ratio <= 512)
+			ratio /= 2;
+		else
 			continue;
 
 		if (ret < savesub) {
@@ -355,11 +356,14 @@ static int fsl_sai_set_bclk(struct snd_soc_dai *dai, bool tx, u32 freq)
 			sai->mclk_id = id;
 			savesub = ret;
 		}
+
+		if (ret == 0)
+			break;
 	}
 
 	if (savediv == 0) {
 		dev_err(dai->dev, "failed to derive required %cx rate: %d\n",
-				tx ? 'T' : 'R', ret);
+				tx ? 'T' : 'R', freq);
 		return -EINVAL;
 	}
 
-- 
1.8.0

